source/fuzz/transformation_replace_copy_memory_with_load_store.cpp

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_replace_copy_memory_with_load_store.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationReplaceCopyMemoryWithLoadStore::
+    TransformationReplaceCopyMemoryWithLoadStore(
+        const spvtools::fuzz::protobufs::
+            TransformationReplaceCopyMemoryWithLoadStore& message)
+    : message_(message) {}
+
+TransformationReplaceCopyMemoryWithLoadStore::
+    TransformationReplaceCopyMemoryWithLoadStore(
+        uint32_t source_value,
+        const protobufs::InstructionDescriptor& instruction_descriptor) {
+  message_.set_source_value(source_value);
+  *message_.mutable_instruction_descriptor() = instruction_descriptor;
+}
+
+bool TransformationReplaceCopyMemoryWithLoadStore::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // |message_.source_value| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.source_value())) {
+    return false;
+  }
+  // The instruction to be replaced must be defined and of opcode
+  // OpCopyMemory.
+  auto copy_memory_instruction =
+      FindInstruction(message_.instruction_descriptor(), ir_context);
+  if (!copy_memory_instruction ||
+      copy_memory_instruction->opcode() != SpvOpCopyMemory) {
+    return false;
+  }
+  // It must be valid to insert the OpStore and OpLoad instruction before it.
+  return (fuzzerutil::CanInsertOpcodeBeforeInstruction(
+              SpvOpStore, copy_memory_instruction) &&
+          fuzzerutil::CanInsertOpcodeBeforeInstruction(
+              SpvOpLoad, copy_memory_instruction));
```

Comment:
I think we can get rid of this.  OpCopyMemory is of the same nature as OpLoad and OpStore, so I believe it's always legitimate to insert the latter in place of the former in terms of where in the module they can appear.

If you can think of an example to the contrary then keep this and add a test for that example that demonstrates that "false" can actually be returned here under the right conditions.  Otherwise drop the code and just return true here (or change the previous if statement to a return).

Reply:
Thanks for pointing this out.